command/kv_rollback.go

```diff
@@ -0,0 +1,258 @@
+		if secret == nil || secret.Data == nil {
+			c.UI.Error(fmt.Sprintf("No value found at %s", path))
+			return 2
+		}
+
+		// Verify metadata found
+		rawMeta, ok := secret.Data["metadata"]
+		if !ok || rawMeta == nil {
+			c.UI.Error(fmt.Sprintf("No metadata found at %s; rollback only works on existing data", path))
+			return 2
+		}
+		meta, ok = rawMeta.(map[string]interface{})
+		if !ok {
+			c.UI.Error(fmt.Sprintf("Metadata found at %s is not the expected type (JSON object)", path))
+			return 2
+		}
+		if meta == nil {
+			c.UI.Error(fmt.Sprintf("No metadata found at %s; rollback only works on existing data", path))
+			return 2
+		}
+
+		// Verify current data found
+		rawData, ok := secret.Data["data"]
+		if !ok || rawData == nil {
+			c.UI.Error(fmt.Sprintf("No data found at %s; rollback only works on existing data", path))
+			return 2
+		}
+		data, ok := rawData.(map[string]interface{})
+		if !ok {
+			c.UI.Error(fmt.Sprintf("Data found at %s is not the expected type (JSON object)", path))
+			return 2
+		}
+		if data == nil {
+			c.UI.Error(fmt.Sprintf("No data found at %s; rollback only works on existing data", path))
+			return 2
+		}
+	}
+
+	casVersion := meta["version"]
+
+	// Set the version parameter
+	versionParam := map[string]string{
+		"version": fmt.Sprintf("%d", c.flagVersion),
+	}
+
+	// Now run it again and read the version we want to roll back to
+	var data map[string]interface{}
+	{
+		secret, err := kvReadRequest(client, path, versionParam)
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Error doing pre-read at %s: %s", path, err))
+			return 2
+		}
+
+		// Make sure a value already exists
+		if secret == nil || secret.Data == nil {
+			c.UI.Error(fmt.Sprintf("No value found at %s", path))
+			return 2
+		}
+
+		// Verify metadata found
+		rawMeta, ok := secret.Data["metadata"]
+		if !ok || rawMeta == nil {
+			c.UI.Error(fmt.Sprintf("No metadata found at %s; rollback only works on existing data", path))
+			return 2
+		}
+		meta, ok := rawMeta.(map[string]interface{})
+		if !ok {
+			c.UI.Error(fmt.Sprintf("Metadata found at %s is not the expected type (JSON object)", path))
+			return 2
+		}
+		if meta == nil {
+			c.UI.Error(fmt.Sprintf("No metadata found at %s; rollback only works on existing data", path))
+			return 2
+		}
+
+		// Verify old data found
+		rawData, ok := secret.Data["data"]
+		if !ok || rawData == nil {
+			c.UI.Error(fmt.Sprintf("No data found at %s; rollback only works on existing data", path))
+			return 2
+		}
+		data, ok = rawData.(map[string]interface{})
+		if !ok {
+			c.UI.Error(fmt.Sprintf("Data found at %s is not the expected type (JSON object)", path))
+			return 2
+		}
+		if data == nil {
+			c.UI.Error(fmt.Sprintf("No data found at %s; rollback only works on existing data", path))
+			return 2
+		}
+
+		if meta["deletion_time"] != nil && meta["deletion_time"].(string) != "" {
+			c.UI.Error(fmt.Sprintf("Cannot roll back to a version that has been deleted"))
```

Comment:
Were you able to test if this check works? I think if the version has been deleted then no data would be returned and the above data checks would fail first.

Reply:
I thought if it was deleted it 404s but still returns `metadata`. I'll check.